improve gitk use of themes

mark987 opened the pull request:
This is addressing #33, and updates gitk to more fully implement themes. Two new configuration variables are added:
* theme          - this is the name of the theme to use (clam, alt, classic, default, ...)
* themeloader - this is the full pathname of a tcl script that when sourced defines one or more themes to be added.

I find some good dark mode themes in the awthemes-10.4.0.zip file available at [tcl-awthemes](https://sourceforge.net/projects/tcl-awthemes/), but most of the themes in that package require Tcl 9.0 (or some additional packages for Tcl 8.6). 

For instance, unzip the above file in /tmp, then edit gitk's config file to have

`set themeloader /tmp/awthemes-10.4.0/awdark.tcl`
`set theme awdark`

This is a good dark mode and works under Tcl 8.6 and Tcl 9.0. Colors must be adjusted in the config file to handle the canvas  objects (foreground and background colors at least).

There is an option in the config page (on the colors page) to select among the installed themes as well.

j6t commented:
Thank you very much, this is very much appreciated!

Here's a bit of feedback after a cursory review.

Did you consider replacing the arrows with Unicode text, ← U+2190, → U+2192, ↑ U+2191, ↓ U+2193. Granted, they are tiny, depending on the font used, but probably still look better than the text. Also, getting the translation of the single words correct that are the button labels is probably at matter of luck.

For the time being, we should not remove the `ui...` color variables from the list of variables that are persisted. Otherwise, we destroy settings for users who switch between this improved series and a version that does not have it. (Of course, we should remove all *uses* of these variables.)

ilyagr commented:
This works well for me on my Mac! Thank you!

I'm not sure whether the dark version of the aqua theme is good for everyone with this PR or whether it still depends on the adjustment I described in https://github.com/j6t/gitk/issues/33#issuecomment-3314531200. But, regardless, the ability to switch the theme to something else is very helpful.

One suggestion: the positioning of the "Apply" button is confusing, since it acts so differently from all the other buttons on that UI. I'd move it to the right of the drop-down, or below it.

<img width="462" height="633" alt="image" src="https://github.com/user-attachments/assets/f06b6f04-75c7-4e47-bced-436359d69930" />

lemzwerg commented:
This is very nice, thanks!  Using commit f5b8b1d7a5f0b2405fa5a1cd6723601550f3fbd7 of the 'j6t-testing' branch the I get the following result.

<img width="1280" height="680" alt="image" src="https://github.com/user-attachments/assets/ee27742c-99ea-4a70-a397-406872227f82" />

@j6t, do you plan to add a dark theme (together with the necessary suite of files) to `gitk`?

j6t commented:
Thank you for your feedback, @lemzwerg , @ilyagr . That's very encouraging.

I do not plan to ship a theme with Gitk.

mark987 commented:
Thanks for the testing.

As a result, I have adjusted the color config page to make the "apply" button's function more clear.

Regarding patches to add a dark theme and/or better instructions:
I gave instructions in the description of this PR for testing, but I would expect instructions on how to use this new capability to be written by someone actually using this to accomplish the purposes in #33. Those need to include not only finding the theme, but also how to adjust all of the colors in the configuration file as many need adjustment. This won't come from me. 

Nothing prevents anyone from offering a patch to include a dark theme with gitk: it needs to be tested on the full range of platforms, not interfere with those happy with the defaults, conform to the gpl, etc. Again, this won't come from me.

A point some have missed: the capability to recolor the gui using the classic widgets was explicitly excluded on Windows: apparently the capability had undesirable side-effects. This new capability is fully cross-platform.

j6t commented:
Thank you, I'll queue this round.

In case someone wants to contribute documentation: the Gitk manual is in the Git repository.

j6t commented:
I've pushed out 659c87180d69e9a2 with some cosmetic fixups. I can squash them in if you agree.

@lemzwerg 
> If I rescale the preferences windows to see longer entries for the 'Theme definition file', the box containing the file name doesn't become wider as I hoped.

This could require more changes and can go in its own commit.

> If I press the 'Choose...' button, the background of the selection window is white.  I would have expected a dark background, too.

This is not something we can solve.

tojamrok commented:
I have an issue with combination of:
- MacBook DPI (raported 144 but actual 220),
- bitmap font,
- tk scaling.

I can hack my way through gitk to make the bitmap font scaled properly with:
- negative font size in `set mainfont` in `~/.config/git/gitk` and
- `tk scaling 3.06` expressed before launching `gitk`, and changing gitk's `parsefont`

j6t commented:
> But then, I wonder whether this merge request would handle my use case in a better way with themes.

Please report your test results.

mark987 commented:
Updated final commit to sort the list of themes presented.

mark987 commented:
I sorted out how to make the config dialog resizeable without major rework. Please try. This is superfluous to the intent of the PR, of course, so can be handled independently, though conflicts occur if trying to apply this without the config dialog changes already in this PR.

ilyagr commented:
Two minor issues with resizing. TBH, if they are hard to fix, they might not be worth fixing.

- the text "The theme definition file may affect all themes" gets cut off instead of wrapped.
- I can resize the dialog to 0 size. No scroll bar or anything appears to move the view if it's too small. (I think the best solution would be to limit how small it could get, but other options exist, including just calling it good enough for now).

j6t commented:
@ilyagr Both issues are in the "if it hurts, don't do it" camp, IMHO.

j6t commented:
I'll fast-track this round of the topic. Thanks again.

mark987 commented:
There is no widget that will dynamically resize and wrap text: ttk::label will wrap on embedded \n, and ::text will dynamically wrap but only onto an already visible line. So, there is no good approach to dynamically wrapping the label.

Setting a minimum size is simple, but must wait for the geometry manager to complete its work. The last patch adds an "update idletasks" command waiting to complete the dialog. I don't know if this would introduce any additional delay in showing the dialog on a very large repo where the background "chew commits" process may take a long time to complete. Certainly, the geometry manager itself must complete is work before the dialog is shown.

I updated the first commit to address the font dialog as well, also added a comment about work I have not done: basically, the original config dialog was on one page, when split into three via a notebook, the code maintained dependence upon a single entry in column 1 that forces indents on all subsequent rows placed using relative geometry. This works for unknown reasons past the next major heading separator, and even onto notebook page two, but fails on page 3. This really should all be changed to use -padx which is the documented way to affect layout like this. But, that's more work and churn and I didn't do it.

j6t commented:
This is now completed. Thank you very much for putting so much effort into this topic, @mark987.
